README.md
# text-time

**text-time** is a tool to help you to parse text that contain time data to seconds

:warning: Node.js >= 14 is required (Open an issue if you require a lower version of Node.js and explain why)

## Motivations

I noticed in my experience that some config contains time in second and like it is an environment variable or a JSON file, I could not write time with this format: `60 * 60 * 24` (means 1 day). This format allows you to easily understand what time it represents, compared to `86400`. I thought it will be very fun if in my configuration, I may write `1 day`. This project is the purpose.

## Usage

```js
const TextTime = require('text-time');

const tt = new TextTime();
console.log(tt('4 months')); // 10368000
```

## Features

- :zap: International ! (Supports :fr: and :uk:)
- change the output format

### API

### `constructor - new TextTime({ output })`

- `output`: By default, the time is returned in second. You can change the unit like you want.

| unit                     | EN                     | FR                                     |
| ------------------------ | ---------------------- | -------------------------------------- |
| second                   | `second`, `seconds`    | `seconde`, `secondes`                  |
| minute                   | `minute`, `minutes`    | `minute`, `minutes`                    |
| hour                     | `hour`, `hours`        | `heure`, `heures`                      |
| day                      | `day`, `days`          | `jour`, `jours`, `journée`, `journées` |
| month                    | `month`, `months`      | `mois`                                 |
| year                     | `year`, `years`        | `an`, `ans`, `année`, `années`         |

### Todo

- [x] Extract all time unit from text
- [x] Handle several output unit
- [ ] Handle writed number with letter (help: https://github.com/finnfiddle/words-to-numbers)


CHANGELOG.md
# Changelog

## 0.0.3 | 2021-04-25

#### Features:

- Allow to return an ISO 8601 string when text contains "dans"

## 0.0.1 | 2021-04-20

#### Features:

- Base code of the project
- Supports EN and FR
- `output` option in the constructor


lib/parser.spec.js
const { expect } = require('chai');
const { extract } = require('./parser');

describe('parser', () => {
  describe('English', () => {
    it('should parse 7 seconds to 7 seconds', () => {
      expect(extract('7 seconds')).to.eq(7);
    });

    it('should parse 63 minutes to 3780 seconds', () => {
      expect(extract('63 minutes')).to.eq(3780);
    });

    it('should parse 4 hours to 14400 seconds', () => {
      expect(extract('4 hours')).to.eq(3600 * 4);
    });

    it('should parse 5 days to 345600 seconds', () => {
      expect(extract('5 days')).to.eq(3600 * 24 * 5);
    });

    it('should parse 4 months to 10368000 seconds', () => {
      expect(extract('4 months')).to.eq(10368000);
    });

    it('should parse 5 years to 4733640000 seconds', () => {
      expect(extract('5 years')).to.eq(157788000);
    });
  });

  describe('French', () => {
    it('should parse 7 secondes to 7 seconds', () => {
      expect(extract('7 secondes')).to.eq(7);
    });

    it('should parse 63 minutes to 3780 seconds', () => {
      expect(extract('63 minutes')).to.eq(3780);
    });

    it('should parse 4 heures to 14400 seconds', () => {
      expect(extract('4 heures')).to.eq(3600 * 4);
    });

    it('should parse 5 jours to 345600 seconds', () => {
      expect(extract('5 jours')).to.eq(3600 * 24 * 5);
    });

    it('should parse 4 mois to 10368000 seconds', () => {
      expect(extract('4 mois')).to.eq(3600 * 24 * 30 * 4);
    });

    it('should parse 5 années to 473364000 seconds', () => {
      expect(extract('5 années')).to.eq(157788000);
    });

    it('should parse 15 ans to 473364000 seconds', () => {
      expect(extract('15 ans')).to.eq(473364000);
    });

    describe('more complexe cases', () => {
      it('should parse 4 jours et 8 heures to 374400', () => {
        expect(extract('4 jours et 8 heures')).to.eq(374400);
      });
    });
  });
});

test/global.spec.js
const { expect } = require('chai');

const TextTime = require('../lib/index');

describe('Global', () => {
  it('should convert 4 hours to second', () => {
    const tt = new TextTime();

    expect(tt('4 hours')).to.eq(3600 * 4);
  });

  it('should convert 4 days to second', () => {
    const tt = new TextTime();

    expect(tt('4 days')).to.eq(3600 * 4 * 24);
  });

  it('should convert 4 days to hour', () => {
    const tt = new TextTime({ output: 'hour' });

    expect(tt('4 days')).to.eq(4 * 24);
  });

  it('should convert 4 days to milliseconds', () => {
    const tt = new TextTime({ output: 'ms' });

    expect(tt('4 days')).to.eq(3600 * 4 * 24 * 1000);
  });
});


test/date/in.spec.js
const { expect } = require('chai');
const Regex = require('../../lib/regex-constants');

const TextTime = require('../../lib/index');

describe('should return an ISO 8601 string if keyword "in" is detected', () => {
  describe('French', () => {
    it('should return an ISO string in 15 days for "dans 15 jours', () => {
      const tt = new TextTime();

      // Compare the date between a range of 5 ms because the output date could differ
      // according the power of the computer
      const expectingA = Date.now() + 3600 * 24 * 15 * 1000 - 5;
      const expectingB = Date.now() + 3600 * 24 * 15 * 1000 + 5;

      const futureDate = tt('dans 15 jours');

      expect(futureDate).to.match(Regex.ISO8601String);
      expect(new Date(futureDate).getTime()).to.greaterThanOrEqual(expectingA);
      expect(new Date(futureDate).getTime()).to.lessThanOrEqual(expectingB);
    });

    it('should return an ISO string in 1 day for "dans 1 jour', () => {
      const tt = new TextTime();

      // Compare the date between a range of 5 ms because the output date could differ
      // according the power of the computer
      const expectingA = Date.now() + 3600 * 24 * 1000 - 5;
      const expectingB = Date.now() + 3600 * 24 * 1000 + 5;

      const futureDate = tt('dans 1 jour');

      expect(futureDate).to.match(Regex.ISO8601String);
      expect(new Date(futureDate).getTime()).to.greaterThanOrEqual(expectingA);
      expect(new Date(futureDate).getTime()).to.lessThanOrEqual(expectingB);
    });
  });
});


lib/parser.js
const regexNumber = /\d+/;

function extract(text) {
  const second = extractSecond(text);
  const minute = extractMinuteInSecond(text);
  const hour = extractHourInSecond(text);
  const day = extractDayInSecond(text);
  const month = extractMonthInSecond(text);
  const year = extractYearInSecond(text);
  
  return year + month + hour + day + minute + second;
}

function extractSecond(text) {
  const regex = /(\d*[ ]?(seconds|secondes))/;
  const res = regex.exec(text);

  if (res?.[0]) {
    const value = regexNumber.exec(res[0]);
    return parseInt(value, 0);
  }

  return 0;
}

function extractMinuteInSecond(text) {
  const regex = /(\d*[ ]?(minutes))/;
  const res = regex.exec(text);

  if (res?.[0]) {
    const value = regexNumber.exec(res[0]);
    return value * 60;
  }

  return 0;
}

function extractHourInSecond(text) {
  const regex = /(\d*[ ]?(hours|heures))/;
  const res = regex.exec(text);

  if (res?.[0]) {
    const value = regexNumber.exec(res[0]);
    return value * 3600;
  }

  return 0;
}

function extractDayInSecond(text) {
  const regex = /(\d*[ ]?(days|jours))*/gm;
  const res = regex.exec(text);

  if (res?.[0]) {
    const value = regexNumber.exec(res[0]);
    return value * 24 * 3600;
  }

  return 0;
}

function extractMonthInSecond(text) {
  const regex = /(\d*[ ]?(months|mois))*/gm;
  const res = regex.exec(text);

  if (res?.[0]) {
    const value = regexNumber.exec(res[0]);
    return value * 24 * 3600 * 30;
  }

  return 0;
}

function extractYearInSecond(text) {
  const regex = /(\d*[ ]?(years|ans|années))*/gm;
  const res = regex.exec(text);

  if (res?.[0]) {
    const value = regexNumber.exec(res[0]);
    return value * 365.25 * 24 * 3600;
  }

  return 0;
}

module.exports = { extract };


lib/index.js
const { extract, detectIfDateIsWanted } = require('./parser');

class TextTime {
  constructor({ output } = { output: 'seconds' }) {
    this.output = output;

    return this.parse.bind(this);
  }

  parse(text) {
    const seconds = extract(text);

    switch (detectIfDateIsWanted(text)) {
    case 'future':
      return new Date(Date.now() + seconds * 1000).toISOString();
    default:
      break;
    }

    switch (this.output) {
    case 'ms': case 'millisecond': case 'milliseconds': case 'milliseconde': case 'millisecondes':
      return seconds * 1000;
    case 'minutes':
      return seconds / 60;
    case 'hour': case 'hours': case 'heure': case 'heures':
      return seconds / 60 / 60;
    case 'day': case 'days': case 'jour': case 'jours': case 'journée': case 'journées':
      return seconds / 60 / 60 / 24;
    case 'year': case 'years': case 'an': case 'ans': case 'année': case 'années':
      return seconds / 60 / 60 / 24 / 30 / 12;
    default:
      return seconds;
    }
  }
}

module.exports = TextTime;
